rio_stac/stac.py

```diff
@@ -53,27 +56,105 @@ def get_metadata(
 
     metadata["bbox"] = list(bbox)
     metadata["footprint"] = bbox_to_geom(bbox)
+    return metadata
+
+
+def get_projection_info(
+    src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT, MemoryFile]
+) -> Dict:
+    """Get projection metadata.
+
+    see: https://github.com/stac-extensions/projection/#item-properties-or-asset-fields
 
-    # Proj
+    """
     try:
         epsg = src_dst.crs.to_epsg() or None
     except AttributeError:
         epsg = None
 
-    metadata["proj"] = {
+    meta = {
         "epsg": epsg,
         "geometry": bbox_to_geom(src_dst.bounds),
         "bbox": list(src_dst.bounds),
         "shape": [src_dst.height, src_dst.width],
         "transform": list(src_dst.transform),
     }
-
     # rasterio can't reproduce wkt2
     # ref: https://github.com/mapbox/rasterio/issues/2044
     # if epsg is None:
-    #      metadata["proj"]["wkt2"] = src_dst.crs.to_wkt()
+    #      meta["wkt2"] = src_dst.crs.to_wkt()
+    return meta
 
-    return metadata
+
+def _get_stats(arr: numpy.ma.array, **kwargs: Any) -> Dict:
+    """Calculate array statistics."""
+    sample, edges = numpy.histogram(arr[~arr.mask])
+    return {
+        "statistics": {
+            "mean": arr.mean().item(),
+            "minimum": arr.min().item(),
+            "maximum": arr.max().item(),
+            "stdev": arr.std().item(),
+            "valid_percent": numpy.count_nonzero(~arr.mask)
+            / float(arr.data.size)
+            * 100,
+        },
+        "histogram": {
+            "count": len(edges),
+            "min": edges.min(),
+            "max": edges.max(),
+            "buckets": sample.tolist(),
+        },
+    }
+
+
+def get_raster_info(
+    src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT, MemoryFile],
+    max_size: int = 1024,
+) -> List[Dict]:
+    """Get raster metadata.
+
+    see: https://github.com/stac-extensions/raster#raster-band-object
+
+    """
+    height = src_dst.height
+    width = src_dst.width
+    if max_size:
+        if max(width, height) > max_size:
+            ratio = height / width
+            if ratio > 1:
+                height = max_size
+                width = math.ceil(height / ratio)
+            else:
+                width = max_size
+                height = math.ceil(width * ratio)
+
+    meta: List[Dict] = []
+
+    area_or_point = src_dst.tags()["AREA_OR_POINT"].lower()
+
+    # Missing `bits_per_sample` and `spatial_resolution`
```

Comment:
I'm not sure how to get `spatial_resolution` 

> Average spatial resolution (in meters) of the pixels in the band.

ok in meters but in which projection?

Reply:
would it be best to use the native coordinate system even if it's not meters?

Reply:
well then we need to update the extension 🤷

Reply:
indeed `spatial_resolution` is not a very well defined. It was added at the very last moment and I don't remember the rationale...
I agree with the idea of refining the spec to link the spatial resolution with the CRS that would then define the unit of the spatial resolution

Reply:
TBH, we already provide the proj extension which has the transform (resolution) and the CRS. IMO the raster extension should not duplicate stuff from the proj extension

Reply:
I agree. I dont remember the reason why we added this field. My only recollection is a discussion about the limitation of the `gsd` field and the need to have a per raster resolution.
@lossyrob @matthewhanson, I remember you were in this discussion, can you enlighten us about the rationale of the `spatial_resolution` field?

Reply:
`spatial_resolution` was added because gsd officially is _sensor_ resolution. So downsampled data like sentinel2 60m still has `10` or `20` for gsd, despite being lower resolution. And you can get it out of the `proj` extension, but only if you have all of proj at your finger tips, so you can resolve whatever projection the data is in.

The discussion around this started here: https://github.com/stac-utils/stactools/pull/88#discussion_r609978081

Reply:
ok then from what I understand of this discussion, there is no requirements to have the spatial resolution only in meters. It is in meters in the case of S2 because of the UTM projection. I only see an issue when it comes to filter items by their `spatial_resolution` in order to have a common unit.

Reply:
int he discussion @cholmes :
> There was debate on if it should be actual pixel sizes, in an array, for times when those might be different. But the conclusion was use the proj extension for that, and this should be a single value in meters, representing the meters per pixel the raster was processed to. So this should be as easy to use as the gsd field. So for the example that this issue started with, we would just replace raster:spatial_resolution for gsd at the asset level, using the 20m and 60m values for the downsampled ones. It could also include the gsd, which would be 10m for all of them (but doing that is optional, since that would be captured at the properties level as well).

Reply:
When we had discussed spatial_resolution, we had talked about it being in `meters` always. I think this is the right choice - this matches what [gsd](https://github.com/radiantearth/stac-spec/blob/master/item-spec/common-metadata.md#instrument) is:
<html>
<body>
<!--StartFragment-->

gsd | number | Ground Sample Distance at the sensor, in meters (m), must be greater than 0.
-- | -- | --

<!--EndFragment-->
</body>
</html>

As Kyle said, spatial_resolution is there to indicate the resolution in meters of the actual raster data, as opposed to GSD, which is the resolution of what the sensor captured (regardless of the raster data, which may have been up- or down-sampled).

Reply:
FYI, This doesn't seems to be well defined for now I'm not going to include by default (at least for now)